### PR TITLE
Refinement rules for composite types; and Split rectification campaign

### PR DESCRIPTION
**Pull request by jonsterling**

### CODE REVIEW

**Comment by favonia** [comment]
Pick one: 🧊 📦 🎁

**Comment by favonia** [comment]
I doubt anyone would actually use these... but it's cool to be complete!

File: src/lib/Semantics.ml [hunk 1]
```diff
@@ -1232,13 +1246,24 @@ and do_el_out con =
...
       cut_frm ~tp ~cut D.KElOut
+    | D.Cut {tp = D.ElCut (D.Split (D.Univ, branches), spine); cut} as con ->
+      let phis, code_clos = List.split branches in
+      let code_fns = List.map (fun clo -> D.Lam (`Anon, clo)) code_clos in
+      splice_tm @@
+      Splice.foreign con @@ fun tm ->
+      Splice.foreign_list (List.map D.cof_to_con phis) @@ fun phis ->
+      Splice.foreign_list code_fns @@ fun code_fns ->
+      Splice.term @@
+      let tp = TB.tp_cof_split @@ List.combine phis @@ List.map (fun code_fn -> TB.el @@ TB.ap code_fn [TB.prf]) code_fns in
```
Comment (jonsterling) [inline src/lib/Semantics.ml:1264]: This isn't quite right yet. It is very challenging!

**Comment by jonsterling** [comment]
This looks promising but I think we should not call that thing `Monoid`, since it has more data than a monoid does.

**Comment by favonia** [comment]
> This looks promising but I think we should not call that thing `Monoid`, since it has more data than a monoid does.

Well, I do not have a better name for this. I think the type of `Monad` polymorphic seq (in previous iteration) are equivalent to the type of all monoids by parametricity (-ish), but the ones with a fixed type `key` I have no ideas.

**Comment by jonsterling** [comment]
Well, we can pick literally any name except Monoid. It is ad hoc, which is OK.

**Comment by jonsterling** [comment]
It probably makes sense to not put it in the Basis, because it is only superficially a general-purpose notion.

**Comment by favonia** [comment]
Okay. Just to clarify, would you be okay with putting Monoid under CoolBasis if `seq` were polymorphic?

**Comment by favonia** [comment]
(Actually, let me postpone the movement and wait until I see the algorithm in your mind.)

**Comment by jonsterling** [comment]
So, let me break away from the discussion of the specific, and say that I feel all this stuff is huge overkill. Let me just give you my pseudocode and you can decide the best way to achieve it.

```ocaml
let cool_quote_con tp con = 
  let* phis = unreduced_cofibrations in
  (* this is the of leaf nodes for the inversion search;
     it would be the singleton [top] if there is no unreduced cofibration *)
  
  let* tbranches =
    phis |> MU.map @@ fun phi ->
    let* tphi = quote_cof phi
    bind_var ~abort:S.CofAbort (D.TpPrf phi) @@ fun prf ->
    let+ tm = uncool_quote_con tp con
    tphi, tm

  let* ttp = quote_tp tp in (* maybe we will need to have `cool_quote_tp` too *)

  ret @@ S.CofSplit (ttp, tbranches)

(* TODO: optimize to not produce a split in the case of [top] *)
```

The above is literally my whole proposal. Does it make sense, or is something missing?

**Comment by jonsterling** [comment]
@favonia To answer your question about monoid and polymorphism, yes that matches my feelings --- if it were polymorphic,  then it would be (up to some laws) just a very Cool version of a monoid.

**Comment by favonia** [comment]
CJHM + CCHM

File: src/lib/Monads.ml [hunk 2]
```diff
@@ -127,60 +130,65 @@ struct
...
+    abort_if_inconsistent abort m
+
+  let restrict (type a) ~(splitter:(D.cof * a m) list -> a m) phis (m : a m) : a m =
+    let module P = struct
+      type t = a m
+      let zero = splitter []
+      let seq f cofs =
+        let l = List.map (fun cof -> cof , f cof) cofs in
+        splitter l
+    end in
```
Comment (favonia) [inline src/lib/Monads.ml:155]: I will probably remove this scary `module` stuff once we have a correct version.

**Comment by favonia** [comment]
@jonsterling I chased down the code and found the "problem" (if we think this is bad): `T.Chk.syn`. It quotes a `dim` variable. Fixing it is easy, but someone should make a wish. There are at least these solutions:

1. Drop unreduced joins when quoting a variable.
2. Special treatment when quoting a dim/cof.
3. Deal with split dims/cofs!

**Comment by favonia** [comment]
Remaining refactoring:

- [ ] Expose `S.dump_tp`
- [ ] Hide the splitter from the `Quote` interface. Instead, expose interfaces that take additional phis.
- [ ] Revamp CofEnv and examine all usages of the API

**Comment by favonia** [comment]
> @jonsterling I chased down the code and found the "problem" (if we think this is bad): `T.Chk.syn`. It quotes a `dim` variable. Fixing this particular tactic is easy, but someone should make a wish. There are at least these solutions:
> 
>     1. Dropping unreduced joins when quoting a variable.
> 
>     2. Special treatment when quoting a dim/cof.
> 
>     3. Dealing with split dims/cofs by having splits everywhere!

The next problematic function is `QTB.lam`. The body might need to be split when the domain is `[some join]`. We need to have the type of the body as well.

File: src/lib/Tactic.ml [hunk 3]
```diff
@@ -62,7 +62,7 @@ struct
   let prf phi = {tp = D.TpPrf phi; con = D.Prf}
   let con {tp; con} = con
   let syn {tp; con} =
-    let+ tm = EM.lift_qu @@ Qu.quote_con tp con in
+    let+ tm = EM.quote_con' tp con in
```
Comment (favonia) [inline src/lib/Tactic.ml:65]: This is the Bug 1. @jonsterling 

Quick Fix 1: `EM.drop_joins_and_lift_qu @@ Qu.quote_con tp con`

File: src/lib/Quote.ml [hunk 4]
```diff
@@ -33,7 +33,7 @@ sig
 end =
 struct
   let lam ?(ident = `Anon) tp mbdy =
-    bind_var ~abort:S.CofAbort tp @@ fun arg ->
+    bind_var ~splitter:(fun _ -> ret S.CofAbort) tp @@ fun arg ->
```
Comment (favonia) [inline src/lib/Quote.ml:36]: Bug 2: `QTB.lam` needs the type of the body for possible splitting.

**Comment by favonia** [comment]
I think we are very closed...

```test/hcom-type.cooltt``` seems to be generating errors? @jonsterling

**Comment by favonia** [comment]
@jonsterling @cangiuli The failure during quotation does not seem to propagate to the outermost layer. The exit code is still 0 despite the errors in `test/hcom-type.cooltt`.

**Comment by jonsterling** [comment]
@favonia I am notseeing any errors in `hcom-type.cooltt`; what's the problem?

**Comment by jonsterling** [comment]
Oh, I do see it if I run from the commandline.

**Comment by favonia** [comment]
> @favonia I am notseeing any errors in `hcom-type.cooltt`; what's the problem?

That's why I'm tagging Carlo. :cq your vim and run it in the command line.

**Comment by favonia** [comment]
Everyone's their own genie!

**Comment by favonia** [comment]
I think you should generalize this and apply it to all the branch-like constructors (that is something we can refactor further).

**Comment by favonia** [comment]
@jonsterling @cangiuli

Let me summarize the situation:

1. It seems beneficial to keep two monads. The one with unreduced joins for normal quoting, and the one without unreduced joins for equality checking. The issue described in #128 is no longer a concern with untyped splitting.
2. The main benefit of eager splitting seems to be fast hcom checking. However, I think this could be achieved by another upgrade of CofEnv (see below).
3. The main drawback of eager splitting as implemented in this PR is that the types of the goals are split "correctly" but will confuse users.
4. This PR has many bug fixes and improvements that should be recycled even if we abandon it.

My proposal:

1. Use only (unsplit) QuM in Quote, sacrificing the current speedup but immediate restoring usability.
2. Merge this PR after Step 1.
3. Have another PR to improve CofEnv so that it will automatically prune and reduce joins. This does not involve any API changes.
4. After Step 3, extend CofEnv with a new set of API to do "shallow" splitting as needed in `D.TpSplit` and `D.Split`, V types, etc. to reduce duplicate code and apply full optimizations in all cases. The shallow splitting API should apply at least these two optimizations: (1) if a branch is always true, then use it directly, otherwise (2) if a branch can never be triggered (as in pattern matching) after the branches before it, then drop it.

**Comment by jonsterling** [comment]
@favonia This is mostly correct, but I need to add that the problem with the eager splitting may not be totally limited to the user experience. One "dark secret" is that the refinement is currently sub-equational, because while `x => M` might succeed on a goal like `(x : A) -> [phi => B | psi | C]`, it would not succeed on the definitionally equivalent goal `[phi => (x : A) -> B | psi => (x : A) -> C]`. But to turn the latter into the former is a unification problem that we do not currently know how to solve in general. So the eager splitting causes information loss.

**Comment by jonsterling** [comment]
@favonia @cangiuli I unofficially approve this PR for merge. GitHub won't let me officially approve it, because I'm the one who opened it LOL

**Comment by favonia** [comment]
@jonsterling Okay. You can overwrite the CI checking, right?

**Pull request merged**